#### About
> Relation Extraction

In the first text pair, the relation mentioned in the second text indicates that "John" is the CEO of "Abc." In this case, we want the model to predict a positive relation between the two texts, so we assign a label of 1 to indicate a positive relation.

In the second text pair, the relation mentioned in the second text indicates that "Apple" is a technology company. Since this relation does not match the first text, we want the model to predict a negative relation, so we assign a label of 0 to indicate a negative relation.

In [15]:
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense


In [16]:
# Manually defined text pairs and labels
text_pairs = [
    ("John is the CEO of Apple.", "John - CEO - Abc"),
    ("Abc is a technology company.", "Abc - company - technology")
]

In [17]:

labels = [1, 0]  # 1 for positive relation, 0 for negative relation


In [18]:
# Tokenize the text pairs
tokens = set()
for text_pair in text_pairs:
    tokens.update(text_pair[0].split())
    tokens.update(text_pair[1].split())


In [19]:
# Create word-to-index mapping
word2idx = {word: idx + 1 for idx, word in enumerate(tokens)}
vocab_size = len(word2idx) + 1


In [20]:

# Convert text pairs to sequences of indices
text_sequences = []
for text_pair in text_pairs:
    sequence = []
    for word in text_pair[0].split():
        sequence.append(word2idx[word])
    text_sequences.append(sequence)


In [21]:
# Pad sequences to the same length
max_length = max(len(sequence) for sequence in text_sequences)
padded_sequences = []
for sequence in text_sequences:
    padded_sequence = np.pad(sequence, (0, max_length - len(sequence)))
    padded_sequences.append(padded_sequence)


In [22]:
# Convert to numpy arrays
X = np.array(padded_sequences)
y = np.array(labels)

In [23]:
# Define the model
embedding_dim = 50
hidden_units = 50

In [24]:

model = Sequential()
model.add(Embedding(vocab_size, embedding_dim, input_length=max_length))
model.add(LSTM(hidden_units))
model.add(Dense(1, activation='sigmoid'))

2023-05-15 18:22:12.766237: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_2_grad/concat/split_2/split_dim' with dtype int32
	 [[{{node gradients/split_2_grad/concat/split_2/split_dim}}]]
2023-05-15 18:22:12.768514: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_grad/concat/split/split_dim' with dtype int32
	 [[{{node gradients/split_grad/concat/split/split_dim}}]]
2023-05-15 18:22:12.770359: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You mus

In [25]:


# Compile and train the model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X, y, epochs=10, batch_size=1)













Epoch 1/10


2023-05-15 18:22:17.732198: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_2_grad/concat/split_2/split_dim' with dtype int32
	 [[{{node gradients/split_2_grad/concat/split_2/split_dim}}]]
2023-05-15 18:22:17.737835: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_grad/concat/split/split_dim' with dtype int32
	 [[{{node gradients/split_grad/concat/split/split_dim}}]]
2023-05-15 18:22:17.740395: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You mus

2/2 [==============================] - 3s 12ms/step - loss: 0.6968 - accuracy: 0.5000
Epoch 2/10
2/2 [==============================] - 0s 14ms/step - loss: 0.6903 - accuracy: 0.5000
Epoch 3/10
2/2 [==============================] - 0s 13ms/step - loss: 0.6856 - accuracy: 0.5000
Epoch 4/10
2/2 [==============================] - 0s 10ms/step - loss: 0.6804 - accuracy: 1.0000
Epoch 5/10
2/2 [==============================] - 0s 12ms/step - loss: 0.6749 - accuracy: 1.0000
Epoch 6/10
2/2 [==============================] - 0s 14ms/step - loss: 0.6688 - accuracy: 1.0000
Epoch 7/10
2/2 [==============================] - 0s 19ms/step - loss: 0.6619 - accuracy: 1.0000
Epoch 8/10
2/2 [==============================] - 0s 14ms/step - loss: 0.6538 - accuracy: 1.0000
Epoch 9/10
2/2 [==============================] - 0s 14ms/step - loss: 0.6443 - accuracy: 1.0000
Epoch 10/10
2/2 [==============================] - 0s 16ms/step - loss: 0.6333 - accuracy: 1.0000


In [26]:
# Use the model for prediction
test_text_pairs = [
    ("John works at Bcd.", "John - employee - Bcd"),
    ("Abc makes smartphones.", "Abc - manufacturer - smartphones")
]


In [27]:

for test_text_pair in test_text_pairs:
    test_sequence = []
    for word in test_text_pair[0].split():
        if word in word2idx:
            test_sequence.append(word2idx[word])
        else:
            test_sequence.append(0)  # Use 0 for out-of-vocabulary words
    test_sequence = np.array(test_sequence)
    test_sequence = np.pad(test_sequence, (0, max_length - len(test_sequence)))
    test_sequence = np.reshape(test_sequence, (1, -1))

    prediction = model.predict(test_sequence)
    print("Text Pair:", test_text_pair)
    print("Prediction:", prediction)

1/1 [==============================] - 0s 473ms/step


2023-05-15 18:22:46.149255: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_2_grad/concat/split_2/split_dim' with dtype int32
	 [[{{node gradients/split_2_grad/concat/split_2/split_dim}}]]
2023-05-15 18:22:46.151201: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_grad/concat/split/split_dim' with dtype int32
	 [[{{node gradients/split_grad/concat/split/split_dim}}]]
2023-05-15 18:22:46.152702: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You mus

Text Pair: ('John works at Bcd.', 'John - employee - Bcd')
Prediction: [[0.45906538]]
1/1 [==============================] - 0s 39ms/step
Text Pair: ('Abc makes smartphones.', 'Abc - manufacturer - smartphones')
Prediction: [[0.4524533]]


Model exhibits 0.45 probability that there is a positive relation